assets/src/components/groupedAutocomplete.tsx

```diff
@@ -0,0 +1,731 @@
+   */
+  setCursorLocation: (forceLocation: CursorState) => void
+}
+
+/**
+ * Helper hook to pre-configure functions with {@link LengthsContext} from the
+ * {@link groups groups parameter} and a few specialty functions.
+ *
+ * @param groups The {@link AutocompleteDataGroup} array used to bounds check
+ * {@link CursorState} arguments and changes.
+ *
+ * @returns The current cursor state and functions to modify the state.
+ */
+const useCursorLocationFromGroups = (
+  groups: AutocompleteDataGroup[],
+  onCursor?: AutocompleteCursorEventProps
+): CursorReducerControls => {
+  const lengthsContext = groups.map(({ group }) => group.options.length)
+
+  const [cursorLocation, dispatchCursorLocation] = useReducer(
+    cursorLocationReducer,
+    undefined
+  )
+
+  const commonArgs = { lengthsContext, onCursor }
+
+  return {
+    /**
+     * The current {@link CursorLocation `cursorLocation`}
+     */
+    cursorLocation,
+    /**
+     * Function to move the {@link CursorLocation `cursorLocation`} relatively
+     * within the {@link LengthsContext}
+     */
+    updateCursorLocation: (action: CursorLocationAction) =>
+      dispatchCursorLocation({ action, ...commonArgs }),
+    /**
+     * Set the {@link CursorLocation}, if the {@link CursorLocation} is valid
+     * within the {@link LengthsContext}.
+     */
+    setCursorLocation: (forceLocation: CursorState) =>
+      dispatchCursorLocation({ forceLocation, ...commonArgs }),
+  }
+}
+// #endregion Cursor Reducer
+
+// #region Autocomplete Control Impl
+/**
+ * Props which expose references for controlling the {@link GroupedAutocomplete}
+ */
+export type GroupedAutocompleteControlRefProps = {
+  /**
+   * Reference object containing functions for controlling the
+   * {@link GroupedAutocomplete}.
+   */
+  controllerRef?: MutableRefObject<GroupedAutocompleteControls | null>
+}
+
+/**
+ * Functions to control the cursor state of the {@link GroupedAutocomplete}
+ */
+export interface GroupedAutocompleteControls {
+  /**
+   * Moves focus to first available option if it exists.
+   */
+  focusCursorToFirstOption: () => void
+  /**
+   * Deletes the cursor in the control, relinquish's focus control.
+   */
+  forgetCursor: () => void
+}
+
+/**
+ * General configuration props for the {@link GroupedAutocomplete} control.
+ */
+export interface GroupedAutocompleteProps
+  extends GroupedAutocompleteControlRefProps {
+  /**
+   * The ID of the `listbox` control.
+   *
+   * Mainly used for aria compatibility.
+   */
+  id?: string
+  /**
+   * The groups and options to render as clickable elements in the control.
+   */
+  optionGroups: AutocompleteDataGroup[]
+  /**
+   * The name of the control. Read to screen readers when control receives focus.
+   */
+  controlName: ReactNode
+  /**
+   * React Component to use as the heading and aria-label for the results list.
+   *
+   * This component must expose the following props
+   * - `aria-hidden`
+   * - `className`
+   * - `id`
```

Comment:
question (non-blocking): Is this needed as part of the props doc? It seems to me that this is taken care of within the component as long as a string value is passed for "Heading"

Reply:
I had this [article](https://dev.to/s_aitchison/psa-stop-hard-coding-heading-levels-in-your-react-components-2ekp) in mind when writing this, in the case that we did end up with "automatic" heading levels in the app, I was thinking it'd be good to callout what interface was required. Though I agree it's not needed given it's simply trying to predict the future.